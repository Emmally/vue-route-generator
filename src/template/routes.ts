import { PageMeta } from '../resolve'

function createChildrenRoute(children: PageMeta[]): string {
  return `,
    children: [${children.map(createRoute).join(',')}]`
}

function createRoute(meta: PageMeta): string {
  const children = !meta.children ? '' : createChildrenRoute(meta.children)

  return `
  {
    name: '${meta.name}',
    path: '${meta.path}',
    component: () => import(/* webpackChunkName: "${meta.name}" */ '${
    meta.component
  }')${children}
  }`
}

export function createRoutes(meta: PageMeta[]): string {
  const code = meta.map(createRoute).join(',')
  return `export default [${code}]`
}
