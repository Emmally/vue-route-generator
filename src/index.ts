import * as fg from 'fast-glob'
import { createRoutes } from './template/routes'
import { resolveRoutePaths } from './resolve'

export interface GenerateConfig {
  pages: string
  importPrefix?: string
}

export function generateRoutes({
  pages,
  importPrefix = '@/pages/'
}: GenerateConfig): string {
  const pagePaths = fg.sync<string>('**/*.vue', {
    cwd: pages,
    onlyFiles: true
  })

  const metaList = resolveRoutePaths(pagePaths, importPrefix)

  return createRoutes(metaList)
}
