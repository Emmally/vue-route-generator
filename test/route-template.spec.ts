import { createRoutes } from '../src/template/routes'
import { PageMeta } from '../src/resolve'

describe('Route template', () => {
  it('should generate routes', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        path: '/foo',
        component: '@/pages/foo.vue'
      },
      {
        name: 'bar',
        path: '/bar',
        component: '@/pages/bar.vue'
      }
    ]

    expect(createRoutes(meta)).toMatchSnapshot()
  })

  it('should generate nested routes', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        path: '/foo',
        component: '@/pages/foo.vue',
        children: [
          {
            name: 'bar',
            path: 'bar',
            component: '@/pages/bar.vue'
          },
          {
            name: 'baz',
            path: 'baz',
            component: '@/pages/baz.vue'
          }
        ]
      }
    ]

    expect(createRoutes(meta)).toMatchSnapshot()
  })
})
