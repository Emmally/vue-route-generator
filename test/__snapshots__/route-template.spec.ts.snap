// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route template should generate nested routes 1`] = `
"export default [
  {
    name: 'foo',
    path: '/foo',
    component: () => import(/* webpackChunkName: \\"foo\\" */ '@/pages/foo.vue'),
    children: [
  {
    name: 'bar',
    path: 'bar',
    component: () => import(/* webpackChunkName: \\"bar\\" */ '@/pages/bar.vue')
  },
  {
    name: 'baz',
    path: 'baz',
    component: () => import(/* webpackChunkName: \\"baz\\" */ '@/pages/baz.vue')
  }]
  }]"
`;

exports[`Route template should generate routes 1`] = `
"export default [
  {
    name: 'foo',
    path: '/foo',
    component: () => import(/* webpackChunkName: \\"foo\\" */ '@/pages/foo.vue')
  },
  {
    name: 'bar',
    path: '/bar',
    component: () => import(/* webpackChunkName: \\"bar\\" */ '@/pages/bar.vue')
  }]"
`;
